In [2]:
import pandas as pd

In [3]:
import glob

# Method 1: Direct approach for your specific files
# Read Plant 1 and Plant 2 Generation Data
plant1_gen = pd.read_csv('Plant_1_Generation_Data.csv')
plant2_gen = pd.read_csv('Plant_2_Generation_Data.csv')

# Union ALL (stack vertically) the generation data
generation_data = pd.concat([plant1_gen, plant2_gen], ignore_index=True)

# Read Plant 1 and Plant 2 Weather Sensor Data
plant1_weather = pd.read_csv('Plant_1_Weather_Sensor_Data.csv')
plant2_weather = pd.read_csv('Plant_2_Weather_Sensor_Data.csv')

# Union ALL (stack vertically) the weather sensor data
weather_sensor_data = pd.concat([plant1_weather, plant2_weather], ignore_index=True)

# Optional: Save to new CSV files if needed
generation_data.to_csv('Combined_Generation_Data.csv', index=False)
weather_sensor_data.to_csv('Combined_Weather_Sensor_Data.csv', index=False)

# Display first few rows to verify
print("Generation Data Combined Shape:", generation_data.shape)
print(generation_data.head())
print("\nWeather Sensor Data Combined Shape:", weather_sensor_data.shape)
print(weather_sensor_data.head())

Generation Data Combined Shape: (136476, 7)
          DATE_TIME  PLANT_ID       SOURCE_KEY  DC_POWER  AC_POWER  \
0  15-05-2020 00:00   4135001  1BY6WEcLGh8j5v7       0.0       0.0   
1  15-05-2020 00:00   4135001  1IF53ai7Xc0U56Y       0.0       0.0   
2  15-05-2020 00:00   4135001  3PZuoBAID5Wc2HD       0.0       0.0   
3  15-05-2020 00:00   4135001  7JYdWkrLSPkdwr4       0.0       0.0   
4  15-05-2020 00:00   4135001  McdE0feGgRqW7Ca       0.0       0.0   

   DAILY_YIELD  TOTAL_YIELD  
0          0.0    6259559.0  
1          0.0    6183645.0  
2          0.0    6987759.0  
3          0.0    7602960.0  
4          0.0    7158964.0  

Weather Sensor Data Combined Shape: (6441, 6)
             DATE_TIME  PLANT_ID       SOURCE_KEY  AMBIENT_TEMPERATURE  \
0  2020-05-15 00:00:00   4135001  HmiyD2TTLFNqkNe            25.184316   
1  2020-05-15 00:15:00   4135001  HmiyD2TTLFNqkNe            25.084589   
2  2020-05-15 00:30:00   4135001  HmiyD2TTLFNqkNe            24.935753   
3  2020-05-1

In [4]:
generation_data = pd.read_csv('Combined_Generation_Data.csv')

In [5]:
weather_data= pd.read_csv('Combined_Weather_Sensor_Data.csv')

In [6]:
generation_data.groupby('PLANT_ID')['SOURCE_KEY'].nunique()

PLANT_ID
4135001    22
4136001    22
Name: SOURCE_KEY, dtype: int64

In [7]:
generation_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136476 entries, 0 to 136475
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   DATE_TIME    136476 non-null  object 
 1   PLANT_ID     136476 non-null  int64  
 2   SOURCE_KEY   136476 non-null  object 
 3   DC_POWER     136476 non-null  float64
 4   AC_POWER     136476 non-null  float64
 5   DAILY_YIELD  136476 non-null  float64
 6   TOTAL_YIELD  136476 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 7.3+ MB


In [8]:
result = generation_data.groupby('PLANT_ID')['SOURCE_KEY'].nunique().reset_index()
print(result)

   PLANT_ID  SOURCE_KEY
0   4135001          22
1   4136001          22


In [9]:
# Let pandas figure out the format for each value
generation_data['DATE_TIME'] = pd.to_datetime(generation_data['DATE_TIME'], format='mixed')

In [10]:
generation_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136476 entries, 0 to 136475
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   DATE_TIME    136476 non-null  datetime64[ns]
 1   PLANT_ID     136476 non-null  int64         
 2   SOURCE_KEY   136476 non-null  object        
 3   DC_POWER     136476 non-null  float64       
 4   AC_POWER     136476 non-null  float64       
 5   DAILY_YIELD  136476 non-null  float64       
 6   TOTAL_YIELD  136476 non-null  float64       
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 7.3+ MB


In [11]:
generation_data['Hour']=generation_data['DATE_TIME'].dt.hour

In [12]:
weather_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6441 entries, 0 to 6440
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   DATE_TIME            6441 non-null   object 
 1   PLANT_ID             6441 non-null   int64  
 2   SOURCE_KEY           6441 non-null   object 
 3   AMBIENT_TEMPERATURE  6441 non-null   float64
 4   MODULE_TEMPERATURE   6441 non-null   float64
 5   IRRADIATION          6441 non-null   float64
dtypes: float64(3), int64(1), object(2)
memory usage: 302.1+ KB


In [13]:
weather_data['IsDaylight']=(weather_data['IRRADIATION']>0).astype(int)

In [14]:
weather_data['Temp_Difference'] = weather_data['MODULE_TEMPERATURE'] - weather_data['AMBIENT_TEMPERATURE']

In [15]:
generation_data.to_csv("Combined_Generation_Data_Another.csv",index=False)

In [16]:
weather_data.to_csv("Combined_Weather_Data_Another.csv",index=False)

In [17]:
result = generation_data.groupby('PLANT_ID')['DATE_TIME'].agg(lambda x: x.dt.date.nunique()).reset_index(name='nr_days')

In [18]:
result

,PLANT_ID,nr_days
0,4135001,34
1,4136001,34


In [19]:
generation_data.head()

,DATE_TIME,PLANT_ID,SOURCE_KEY,DC_POWER,AC_POWER,DAILY_YIELD,TOTAL_YIELD,Hour
0,2020-05-15,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0,0
1,2020-05-15,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0,0
2,2020-05-15,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0,0
3,2020-05-15,4135001,7JYdWkrLSPkdwr4,0.0,0.0,0.0,7602960.0,0
4,2020-05-15,4135001,McdE0feGgRqW7Ca,0.0,0.0,0.0,7158964.0,0


In [20]:
generation_data.groupby('PLANT_ID')['DATE_TIME'].nunique()

PLANT_ID
4135001    3158
4136001    3259
Name: DATE_TIME, dtype: int64

In [21]:
generation_data['PLANT_ID'].unique()

array([4135001, 4136001])

<!DOCTYPE html>
<html>
<head>
<style>
  body {
    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
    background: #f0f4f8;
    padding: 40px;
  }
  
  table {
    width: 100%;
    border-collapse: collapse;
  }
  
  td {
    border-radius: 20px;
  }
</style>
</head>
<body>

<!-- Option 2: Light Blue with Dark Text -->
<table style="width: 100%;">
  <tr>
    <td style="background: linear-gradient(120deg, #89f7fe 0%, #66a6ff 100%);
               padding: 50px 45px;
               border-radius: 20px;
                box-shadow: 0 10px 30px rgba(0,0,0,0.1);">
      <div style="font-size: 28px; 
                  font-weight: 800; 
                  margin-bottom: 20px;
                  color: #0c4e6e;">
        📊 Inverter Efficiency
      </div>
    </td>
  </tr>
</table>

</body>
</html>              

In [22]:
result = generation_data.groupby('PLANT_ID')[['DC_POWER', 'AC_POWER']].mean().reset_index()
result.columns = ['plant_id', 'avg_dc_power', 'avg_ac_power']

print(result)

   plant_id  avg_dc_power  avg_ac_power
0   4135001   3147.426211    307.802752
1   4136001    246.701961    241.277825


In [23]:
# Calculate means first, then divide
avg_power = generation_data.groupby('PLANT_ID')[['DC_POWER', 'AC_POWER']].mean()
inv_efficiency = (avg_power['AC_POWER'] / avg_power['DC_POWER']) * 100

print(inv_efficiency)

PLANT_ID
4135001     9.779507
4136001    97.801341
dtype: float64


<!DOCTYPE html>
<html>
<head>
<style>
  body {
    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
    background: #f0f4f8;
    padding: 40px;
  }
  
  table {
    width: 100%;
    border-collapse: collapse;
  }
  
  td {
    border-radius: 20px;
  }
</style>
</head>
<body>

<!-- Option 2: Light Blue with Dark Text -->
<table style="width: 100%;">
  <tr>
    <td style="background: linear-gradient(120deg, #89f7fe 0%, #66a6ff 100%);
               padding: 50px 45px;
               border-radius: 20px;
                box-shadow: 0 10px 30px rgba(0,0,0,0.1);">
      <div style="font-size: 28px; 
                  font-weight: 800; 
                  margin-bottom: 20px;
                  color: #0c4e6e;">
        📊 Oneof the two plants (plant_id 4135001) has a DC Power that is 10x higher
than the other. The most reasonable explanation is some kind of error in
the recording or reporting of the data.
      </div>
    </td>
  </tr>
</table>

</body>
</html>         

In [24]:
generation_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136476 entries, 0 to 136475
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   DATE_TIME    136476 non-null  datetime64[ns]
 1   PLANT_ID     136476 non-null  int64         
 2   SOURCE_KEY   136476 non-null  object        
 3   DC_POWER     136476 non-null  float64       
 4   AC_POWER     136476 non-null  float64       
 5   DAILY_YIELD  136476 non-null  float64       
 6   TOTAL_YIELD  136476 non-null  float64       
 7   Hour         136476 non-null  int32         
dtypes: datetime64[ns](1), float64(4), int32(1), int64(1), object(1)
memory usage: 7.8+ MB


In [25]:
import numpy as np

# Filter for specific plant
plant_data = generation_data[generation_data['PLANT_ID'] == 4136001]

# Extract hour from datetime
plant_data['hour_of_day'] = pd.to_datetime(plant_data['DATE_TIME']).dt.hour

# Create temporary columns with zeros replaced by NaN for calculations
plant_data['DC_POWER_clean'] = plant_data['DC_POWER'].replace(0, np.nan)
plant_data['AC_POWER_clean'] = plant_data['AC_POWER'].replace(0, np.nan)

# Calculate hourly aggregates
avg_dc_power_hourly=plant_data.groupby('hour_of_day')['DC_POWER'].mean()
avg_ac_power_hourly=plant_data.groupby('hour_of_day')['AC_POWER'].mean()
avg_dc_clean=plant_data.groupby('hour_of_day')['DC_POWER_clean'].mean()
avg_ac_clean=plant_data.groupby('hour_of_day')['AC_POWER_clean'].mean()

# Calculate efficiency using clean averages (excluding zeros))
inv_efficiency = round((avg_ac_clean / avg_dc_clean) * 100, 2).fillna(0)

# Display efficiency by hour
print(inv_efficiency)


hourly_effi_result = pd.DataFrame({
    'avg_dc_power': avg_dc_power_hourly.values,
    'avg_ac_power': avg_ac_power_hourly.values,
    'pct_inverter_efficiency': round((avg_ac_clean / avg_dc_clean) * 100, 2).fillna(0)
}, index=avg_dc_clean.index).reset_index()[['hour_of_day', 'avg_dc_power', 'avg_ac_power', 'pct_inverter_efficiency']]

hourly_effi_result = hourly_effi_result.sort_values('hour_of_day')

print(hourly_effi_result)

hour_of_day
0      0.00
1      0.00
2      0.00
3      0.00
4      0.00
5     96.59
6     97.14
7     98.03
8     98.03
9     97.86
10    97.73
11    97.66
12    97.64
13    97.67
14    97.78
15    97.92
16    98.04
17    97.93
18    96.94
19     0.00
20     0.00
21     0.00
22     0.00
23     0.00
dtype: float64
    hour_of_day  avg_dc_power  avg_ac_power  pct_inverter_efficiency
0             0      0.000000      0.000000                     0.00
1             1      0.000000      0.000000                     0.00
2             2      0.000000      0.000000                     0.00
3             3      0.000000      0.000000                     0.00
4             4      0.000000      0.000000                     0.00
5             5      0.003212      0.003102                    96.59
6             6     58.695532     57.017699                    97.14
7             7    244.283851    239.474513                    98.03
8             8    507.486537    497.476125                    9

C:\Users\Manar\AppData\Local\Temp\ipykernel_18240\2466724395.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  plant_data['hour_of_day'] = pd.to_datetime(plant_data['DATE_TIME']).dt.hour
C:\Users\Manar\AppData\Local\Temp\ipykernel_18240\2466724395.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  plant_data['DC_POWER_clean'] = plant_data['DC_POWER'].replace(0, np.nan)
C:\Users\Manar\AppData\Local\Temp\ipykernel_18240\2466724395.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a 

## ✅ *"Inverter efficiency peaks at 98.04% during medium-load conditions (244–507 kW DC), which occur in the early morning (hour 7-8) and late afternoon (hour 16)."*

✅ "At maximum power generation (hour 12, 675 kW DC), inverter efficiency is slightly lower at 97.64% due to increased thermal and switching losses."

✅ "At very low power levels (hour 5 and hour 18, <60 kW DC), efficiency drops to 96.6–97.1% due to the impact of fixed standby losses."

✅ "Nighttime hours produce zero power, and inverter efficiency is undefined/zero — not highest.

In [26]:
generation_data['SOURCE_KEY'].nunique()

44

In [27]:
weather_data['SOURCE_KEY'].nunique()

2

## There are 44 unique source keys in the first table and only 2 in the second.

In [28]:
import pandas as pd

# Load data from CSV files
Plant_1_Generation_Data = pd.read_csv('Plant_1_Generation_Data.csv')
Plant_2_Generation_Data = pd.read_csv('Plant_2_Generation_Data.csv')

# Then run your code
plant1_data = Plant_1_Generation_Data.copy()
plant1_data['PLANT_NR'] = 'plant_1'
plant1_data['DC_POWER'] = plant1_data['DC_POWER'] / 10

plant2_data = Plant_2_Generation_Data.copy()
plant2_data['PLANT_NR'] = 'plant_2'

generation_data_clean = pd.concat([plant1_data, plant2_data], ignore_index=True)

print(generation_data_clean.head())
print(f"\nTotal rows: {len(generation_data_clean)}")

          DATE_TIME  PLANT_ID       SOURCE_KEY  DC_POWER  AC_POWER  \
0  15-05-2020 00:00   4135001  1BY6WEcLGh8j5v7       0.0       0.0   
1  15-05-2020 00:00   4135001  1IF53ai7Xc0U56Y       0.0       0.0   
2  15-05-2020 00:00   4135001  3PZuoBAID5Wc2HD       0.0       0.0   
3  15-05-2020 00:00   4135001  7JYdWkrLSPkdwr4       0.0       0.0   
4  15-05-2020 00:00   4135001  McdE0feGgRqW7Ca       0.0       0.0   

   DAILY_YIELD  TOTAL_YIELD PLANT_NR  
0          0.0    6259559.0  plant_1  
1          0.0    6183645.0  plant_1  
2          0.0    6987759.0  plant_1  
3          0.0    7602960.0  plant_1  
4          0.0    7158964.0  plant_1  

Total rows: 136476


In [29]:
import pandas as pd

# Extract hour
weather_sensor_data['hour_of_day'] = pd.to_datetime(weather_sensor_data['DATE_TIME']).dt.hour

# Group by hour and calculate means
result = weather_sensor_data.groupby('hour_of_day')[['AMBIENT_TEMPERATURE', 'MODULE_TEMPERATURE', 'IRRADIATION']].mean().reset_index()

# Add temperature difference
result['avg_temp_diff'] = result['MODULE_TEMPERATURE'] - result['AMBIENT_TEMPERATURE']

# Rename columns and round values
result.columns = ['hour_of_day', 'avg_amb_temp', 'avg_mod_temp', 'avg_irradiation', 'avg_temp_diff']
result = result.round(2)

# Sort by hour
result = result.sort_values('hour_of_day')

print(result)

    hour_of_day  avg_amb_temp  avg_mod_temp  avg_irradiation  avg_temp_diff
0             0         24.01         22.44             0.00          -1.56
1             1         23.73         22.22             0.00          -1.51
2             2         23.50         22.03             0.00          -1.47
3             3         23.32         21.85             0.00          -1.47
4             4         23.15         21.68             0.00          -1.47
5             5         23.03         21.57             0.00          -1.46
6             6         23.23         22.56             0.04          -0.67
7             7         24.42         27.76             0.19           3.33
8             8         25.98         35.35             0.38           9.36
9             9         27.55         41.81             0.55          14.26
10           10         28.91         46.37             0.67          17.46
11           11         30.12         50.08             0.75          19.96
12          

In [30]:
import pandas as pd

# Extract hour
weather_sensor_data['hour_of_day'] = pd.to_datetime(weather_sensor_data['DATE_TIME']).dt.hour

# Group by hour and calculate means
result = weather_sensor_data.groupby('hour_of_day')[['AMBIENT_TEMPERATURE', 'MODULE_TEMPERATURE', 'IRRADIATION']].mean().reset_index()

# Add temperature difference
result['avg_temp_diff'] = result['MODULE_TEMPERATURE'] - result['AMBIENT_TEMPERATURE']

# Rename columns and round values
result.columns = ['hour_of_day', 'avg_amb_temp', 'avg_mod_temp', 'avg_irradiation', 'avg_temp_diff']
result = result.round(2)

# Sort by hour
result = result.sort_values('hour_of_day')

# Display the table
print(result)

# Save to Excel file (will save in the same directory as your script)
excel_filename = 'hourly_weather_data.xlsx'
result.to_excel(excel_filename, index=False, sheet_name='Hourly Weather Data')

print(f"\n✅ Data successfully saved to '{excel_filename}'")
print(f"📍 File location: {os.path.abspath(excel_filename)}")

    hour_of_day  avg_amb_temp  avg_mod_temp  avg_irradiation  avg_temp_diff
0             0         24.01         22.44             0.00          -1.56
1             1         23.73         22.22             0.00          -1.51
2             2         23.50         22.03             0.00          -1.47
3             3         23.32         21.85             0.00          -1.47
4             4         23.15         21.68             0.00          -1.47
5             5         23.03         21.57             0.00          -1.46
6             6         23.23         22.56             0.04          -0.67
7             7         24.42         27.76             0.19           3.33
8             8         25.98         35.35             0.38           9.36
9             9         27.55         41.81             0.55          14.26
10           10         28.91         46.37             0.67          17.46
11           11         30.12         50.08             0.75          19.96
12          

NameError: name 'os' is not defined

## As onewould expect, irradiation is higher during the daytime and near
zero at nighttime. Also, it is interesting to note that the module
temperature is greater than the ambient at daytime and viceversa at
nighttime.


In [73]:
generation_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136476 entries, 0 to 136475
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   DATE_TIME    136476 non-null  object 
 1   PLANT_ID     136476 non-null  int64  
 2   SOURCE_KEY   136476 non-null  object 
 3   DC_POWER     136476 non-null  float64
 4   AC_POWER     136476 non-null  float64
 5   DAILY_YIELD  136476 non-null  float64
 6   TOTAL_YIELD  136476 non-null  float64
 7   PLANT_NR     136476 non-null  object 
dtypes: float64(4), int64(1), object(3)
memory usage: 8.3+ MB


In [74]:
generation_data_clean['DATE_TIME']=pd.to_datetime(generation_data_clean['DATE_TIME'],format='mixed')

In [75]:
# Calculate means first, then divide
avg_power = generation_data.groupby('PLANT_ID')[['DC_POWER', 'AC_POWER']].mean()
inv_efficiency = (avg_power['AC_POWER'] / avg_power['DC_POWER']) * 100

print(inv_efficiency)

PLANT_ID
4135001     9.779507
4136001    97.801341
dtype: float64


In [76]:
import pandas as pd

# Convert DATE_TIME to datetime format (add this at the beginning)
generation_data_clean['DATE_TIME'] = pd.to_datetime(generation_data_clean['DATE_TIME'])
weather_sensor_data['DATE_TIME'] = pd.to_datetime(weather_sensor_data['DATE_TIME'])

# Now your existing code will work
generation_data_clean['HOUR'] = generation_data_clean['DATE_TIME'].dt.hour
avg_power_clean = generation_data_clean.groupby('PLANT_ID')[['DC_POWER', 'AC_POWER']].mean()
gen_efficiency_clean = (avg_power_clean['AC_POWER'] / avg_power_clean['DC_POWER']) * 100

# Part 1: Generation data by PLANT_ID and hour (matching SQL subquery 'a')
gen_hourly = generation_data_clean.groupby(['PLANT_ID', 'HOUR']).agg(
    avg_dc_power=('DC_POWER', 'mean'),
    avg_ac_power=('AC_POWER', 'mean')
).reset_index()

# Calculate efficiency by hour (excluding zeros)
def calc_efficiency(group):
    dc_clean = group[group['DC_POWER'] != 0]['DC_POWER'].mean()
    ac_clean = group[group['AC_POWER'] != 0]['AC_POWER'].mean()
    if dc_clean and dc_clean != 0:
        return round((ac_clean / dc_clean) * 100, 2)
    return 0

gen_hourly['inverter_efficiency'] = generation_data_clean.groupby(['PLANT_ID', 'HOUR']).apply(calc_efficiency).values

# Part 2: Weather data by PLANT_ID and hour (matching SQL subquery 'b')
weather_sensor_data['HOUR'] = weather_sensor_data['DATE_TIME'].dt.hour

weather_hourly = weather_sensor_data.groupby(['PLANT_ID', 'HOUR']).agg(
    avg_amb_temp=('AMBIENT_TEMPERATURE', 'mean'),
    avg_mod_temp=('MODULE_TEMPERATURE', 'mean'),
    avg_irradiation=('IRRADIATION', 'mean')
).reset_index()

# Add temperature difference
weather_hourly['avg_temp_diff'] = weather_hourly['avg_mod_temp'] - weather_hourly['avg_amb_temp']

# Part 3: LEFT JOIN (same as SQL)
final_result = pd.merge(
    gen_hourly,
    weather_hourly,
    on=['PLANT_ID', 'HOUR'],
    how='left'
)

# Round numeric columns
final_result = final_result.round(2)

# Order by PLANT_ID and HOUR (same as SQL ORDER BY)
final_result = final_result.sort_values(['PLANT_ID', 'HOUR']).reset_index(drop=True)

# Rename HOUR to hour_of_day to match SQL output
final_result = final_result.rename(columns={'HOUR': 'hour_of_day'})

print(final_result)

    PLANT_ID  hour_of_day  avg_dc_power  avg_ac_power  inverter_efficiency  \
0    4135001            0          0.00          0.00                  NaN   
1    4135001            1          0.00          0.00                  NaN   
2    4135001            2          0.00          0.00                  NaN   
3    4135001            3          0.00          0.00                  NaN   
4    4135001            4          0.00          0.00                  NaN   
5    4135001            5          0.00          0.00                  NaN   
6    4135001            6         57.81         56.14                97.10   
7    4135001            7        255.14        250.24                98.08   
8    4135001            8        508.86        498.91                98.04   
9    4135001            9        724.74        709.35                97.88   
10   4135001           10        872.04        852.33                97.74   
11   4135001           11        980.56        957.69           

C:\Users\Manar\AppData\Local\Temp\ipykernel_2164\4032249169.py:26: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  gen_hourly['inverter_efficiency'] = generation_data_clean.groupby(['PLANT_ID', 'HOUR']).apply(calc_efficiency).values


In [34]:
import pandas as pd
import os

# Convert DATE_TIME to datetime format - YOUR DATA USES DAY-MONTH-YEAR format
# Format: "15-05-2020 00:00" is day-month-year
generation_data_clean['DATE_TIME'] = pd.to_datetime(generation_data_clean['DATE_TIME'], format='%d-%m-%Y %H:%M')
weather_data['DATE_TIME'] = pd.to_datetime(weather_data['DATE_TIME'], format='%d-%m-%Y %H:%M')

# Now your existing code will work
generation_data_clean['HOUR'] = generation_data_clean['DATE_TIME'].dt.hour
avg_power_clean = generation_data_clean.groupby('PLANT_ID')[['DC_POWER', 'AC_POWER']].mean()
gen_efficiency_clean = (avg_power_clean['AC_POWER'] / avg_power_clean['DC_POWER']) * 100

# Part 1: Generation data by PLANT_ID and hour (matching SQL subquery 'a')
gen_hourly = generation_data_clean.groupby(['PLANT_ID', 'HOUR']).agg(
    avg_dc_power=('DC_POWER', 'mean'),
    avg_ac_power=('AC_POWER', 'mean')
).reset_index()

# Calculate efficiency by hour (excluding zeros)
def calc_efficiency(group):
    dc_clean = group[group['DC_POWER'] != 0]['DC_POWER'].mean()
    ac_clean = group[group['AC_POWER'] != 0]['AC_POWER'].mean()
    if dc_clean and dc_clean != 0:
        return round((ac_clean / dc_clean) * 100, 2)
    return 0

gen_hourly['inverter_efficiency'] = generation_data_clean.groupby(['PLANT_ID', 'HOUR']).apply(calc_efficiency).values

# Part 2: Weather data by PLANT_ID and hour (matching SQL subquery 'b')
weather_data['HOUR'] = weather_data['DATE_TIME'].dt.hour

weather_hourly = weather_data.groupby(['PLANT_ID', 'HOUR']).agg(
    avg_amb_temp=('AMBIENT_TEMPERATURE', 'mean'),
    avg_mod_temp=('MODULE_TEMPERATURE', 'mean'),
    avg_irradiation=('IRRADIATION', 'mean')
).reset_index()

# Add temperature difference
weather_hourly['avg_temp_diff'] = weather_hourly['avg_mod_temp'] - weather_hourly['avg_amb_temp']

# Part 3: LEFT JOIN (same as SQL)
final_result = pd.merge(
    gen_hourly,
    weather_hourly,
    on=['PLANT_ID', 'HOUR'],
    how='left'
)
# Round numeric columns
final_result = final_result.round(2)

# Order by PLANT_ID and HOUR (same as SQL ORDER BY)
final_result = final_result.sort_values(['PLANT_ID', 'HOUR']).reset_index(drop=True)

# Rename HOUR to hour_of_day to match SQL output
final_result = final_result.rename(columns={'HOUR': 'hour_of_day'})

# Display the result
print(final_result)

# Save to Excel file
excel_filename = 'plant_hourly_data.xlsx'

# Create Excel file with formatted table
with pd.ExcelWriter(excel_filename, engine='openpyxl') as writer:
    final_result.to_excel(writer, index=False, sheet_name='Plant Hourly Data')
    
    # Auto-adjust column widths
    worksheet = writer.sheets['Plant Hourly Data']
    for column in worksheet.columns:
        max_length = 0
        column_letter = column[0].column_letter
        for cell in column:
            try:
                if len(str(cell.value)) > max_length:
                    max_length = len(str(cell.value))
            except:
                pass
        adjusted_width = min(max_length + 2, 30)
        worksheet.column_dimensions[column_letter].width = adjusted_width

print(f"\n✅ Data successfully saved to '{excel_filename}'")
print(f"📊 You can now open this file in Microsoft Excel")
print(f"📍 File location: {os.path.abspath(excel_filename)}")

ValueError: time data "2020-05-15 00:00:00" doesn't match format "%d-%m-%Y %H:%M", at position 3158. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

In [82]:
weather_data['DATE_TIME']

0       2020-05-15 00:00:00
1       2020-05-15 00:15:00
2       2020-05-15 00:30:00
3       2020-05-15 00:45:00
4       2020-05-15 01:00:00
               ...         
6436    2020-06-17 22:45:00
6437    2020-06-17 23:00:00
6438    2020-06-17 23:15:00
6439    2020-06-17 23:30:00
6440    2020-06-17 23:45:00
Name: DATE_TIME, Length: 6441, dtype: object

In [48]:
generation_data.to_csv("Combined_Generation_Data.csv",index=False)

In [49]:
weather_data.to_csv("Combines_Weather_Sensor.csv",index=False)